lib/modules.nix

```diff
@@ -87,31 +90,70 @@ rec {
- filterModules = modulesPath: modules:
-   let
-     moduleKey = m: if isString m then toString modulesPath + "/" + m else toString m;
-     disabledKeys = map moduleKey (concatMap (m: m.disabledModules) modules);
-   in
-     filter (m: !(elem m.key disabledKeys)) modules;
-
-  /* Close a set of modules under the ‘imports’ relation. */
-  closeModules = modules: args:
-    let
-      toClosureList = file: parentKey: imap1 (n: x:
-        if isAttrs x || isFunction x then
-          let key = "${parentKey}:anon-${toString n}"; in
-          unifyModuleSyntax file key (applyIfFunction key x args)
-        else
-          let file = toString x; key = toString x; in
-          unifyModuleSyntax file key (applyIfFunction key (import x) args));
-    in
-      builtins.genericClosure {
-        startSet = toClosureList unknownModule "" modules;
-        operator = m: toClosureList m._file m.key m.imports;
-      };
+      # Collects all modules recursively into the form
+      #
+      #   {
+      #     disabled = [ <list of disabled modules> ];
+      #     keys.key1 = {
+      #       module = <module for key1>;
+      #       keys.key1-2 = {
+      #         module = <module for key1-2>;
+      #         keys = ...;
+      #       };
+      #     };
+      #   }
+      #
+      # - `disabled` contains the list of all disabled modules
+      # - `keys` is an attribute containing the module at each key along with all
+      #   modules keys that were imported by that module
+      collectStructuredModules =
+        let
+          collectResults = list: {
+            disabled = concatLists (catAttrs "disabled" list);
+            keys = listToAttrs list;
+          };
+        in parentFile: parentKey: modules: args: collectResults (imap1 (n: x:
+          let
+            module = loadModule args parentFile "${parentKey}:anon-${toString n}" x;
+            collectedImports = collectStructuredModules module._file module.key module.imports args;
+          in {
+            name = module.key;
+            disabled = module.disabledModules ++ collectedImports.disabled;
+            value = { module = module; keys = collectedImports.keys; };
+          }) modules);
+
+      # filterModules :: String -> { disabled, keys } -> [ Module ]
+      #
+      # Filters a structure as emitted by collectStructuredModules by removing all disabled
+      # modules recursively. It returns the final list of unique-by-key modules
+      filterModules = modulesPath: { disabled, keys }:
+        let
+          moduleKey = m: if isString m then toString modulesPath + "/" + m else toString m;
+          disabledKeys = map moduleKey disabled;
+          # Takes a recursive `keys` set as described in collectStructuredModules and flattens
+          # it out to a direct `key -> module` mapping while also filtering disabled
+          # keys
+          flattenKeys = keys:
+            let included = filterAttrs (key: value: ! elem key disabledKeys) keys;
```

Comment:
Did you try `builtins.removeAttrs`? Unlike `filterAttrs`, it's a primop.

Reply:
Ah nice, didn't even occur to me